pkg/restore/restore.go

```diff
@@ -1321,37 +1327,44 @@ func (ctx *restoreContext) restoreItem(obj *unstructured.Unstructured, groupReso
 	return warnings, errs
 }
 
-func isAlreadyExistsError(ctx *restoreContext, obj *unstructured.Unstructured, err error) bool {
+func isAlreadyExistsError(ctx *restoreContext, obj *unstructured.Unstructured, err error, client client.Dynamic) (bool, error) {
 	if err == nil {
-		return false
+		return false, nil
 	}
 	if apierrors.IsAlreadyExists(err) {
-		return true
+		return true, nil
 	}
 	// The "invalid value error" or "internal error" rather than "already exists" error returns when restoring nodePort service in the following two cases:
 	// 1. For NodePort service, the service has nodePort preservation while the same nodePort service already exists. - Get invalid value error
 	// 2. For LoadBalancer service, the "healthCheckNodePort" already exists. - Get internal error
 	// If this is the case, the function returns true to avoid reporting error.
 	// Refer to https://github.com/vmware-tanzu/velero/issues/2308 for more details
 	if obj.GetKind() != "Service" {
-		return false
+		return false, nil
 	}
 	statusErr, ok := err.(*apierrors.StatusError)
 	if !ok || statusErr.Status().Details == nil || len(statusErr.Status().Details.Causes) == 0 {
-		return false
+		return false, nil
 	}
 	// make sure all the causes are "port allocated" error
-	isAllocatedErr := true
 	for _, cause := range statusErr.Status().Details.Causes {
 		if !strings.Contains(cause.Message, "provided port is already allocated") {
-			isAllocatedErr = false
-			break
+			return false, nil
 		}
 	}
-	if isAllocatedErr {
-		ctx.log.Infof("ignore the provided port is already allocated error for service %s", kube.NamespaceAndName(obj))
+
+	// the "already allocated" error may caused by other services, check whether the expected service exists or not
+	if _, err = client.Get(obj.GetName(), metav1.GetOptions{}); err != nil {
+		if apierrors.IsNotFound(err) {
+			ctx.log.Debugf("Service %s not found", kube.NamespaceAndName(obj))
+			return false, nil
+		}
+		return false, errors.Wrapf(err, "Unable to get the service %s", kube.NamespaceAndName(obj))
 	}
-	return isAllocatedErr
+
+	// although the service exists, the NodePort may allocated to others, if this is the case, a warning message will be appended in the restore logs which is handled by the main restore process
```

Comment:
wrong comment?

Reply:
If there are already two services created in the cluster, one is named `svc-1` with NodePort `30001` allocated and the other one is named `svc-2` with NodePort `30002` allocated. When restoring the `svc-2` with NodePort `30001` specified, we'll get the `NodePort already allocated` error. In this case, the restore is succeed with warnings(the warning indicates that the service already exists but it's different with the expected one)